tfhe/src/integer/gpu/key_switching_key/mod.rs

```diff
@@ -0,0 +1,57 @@
+use crate::core_crypto::gpu::lwe_keyswitch_key::CudaLweKeyswitchKey;
+use crate::core_crypto::gpu::CudaStreams;
+use crate::integer::client_key::secret_encryption_key::SecretEncryptionKeyView;
+use crate::integer::gpu::CudaServerKey;
+use crate::integer::ClientKey;
+use crate::shortint::engine::ShortintEngine;
+use crate::shortint::parameters::ShortintKeySwitchingParameters;
+
+#[allow(dead_code)]
+pub struct CudaKeySwitchingKey<'keys> {
+    pub(crate) key_switching_key: CudaLweKeyswitchKey<u64>,
+    pub(crate) dest_server_key: &'keys CudaServerKey,
+}
+
+impl<'keys> CudaKeySwitchingKey<'keys> {
+    pub fn new<'input_key, InputEncryptionKey>(
+        input_key_pair: (InputEncryptionKey, Option<&'keys CudaServerKey>),
+        output_key_pair: (&'keys ClientKey, &'keys CudaServerKey),
+        params: ShortintKeySwitchingParameters,
+        streams: &CudaStreams,
+    ) -> Self
+    where
+        InputEncryptionKey: Into<SecretEncryptionKeyView<'input_key>>,
+    {
+        let input_secret_key: SecretEncryptionKeyView<'_> = input_key_pair.0.into();
+
+        // Creation of the key switching key
+        let key_switching_key = ShortintEngine::with_thread_local_mut(|engine| {
+            engine.new_key_switching_key(&input_secret_key.key, output_key_pair.0.as_ref(), params)
+        });
+        let d_key_switching_key =
+            CudaLweKeyswitchKey::from_lwe_keyswitch_key(&key_switching_key, streams);
+        let full_message_modulus_input =
+            input_secret_key.key.carry_modulus.0 * input_secret_key.key.message_modulus.0;
+        let full_message_modulus_output = output_key_pair.0.key.parameters.carry_modulus().0
+            * output_key_pair.0.key.parameters.message_modulus().0;
+        assert!(
+            full_message_modulus_input.is_power_of_two()
+                && full_message_modulus_output.is_power_of_two(),
+            "Cannot create casting key if the full messages moduli are not a power of 2"
+        );
+        if full_message_modulus_input > full_message_modulus_output {
+            assert!(
+                input_key_pair.1.is_some(),
+                "Trying to build a shortint::KeySwitchingKey \
+                going from a large modulus {full_message_modulus_input} \
+                to a smaller modulus {full_message_modulus_output} \
+                without providing a source ServerKey, this is not supported"
```

Comment:
Do we support this even if a CudaServerKey is provided? In the message, it should maybe be CudaServerKey instead of ServerKey?

Reply:
Sorry, I copy-pasted that method from the CPU implementation. In this flavor, a `CudaServerKey` must be provided, not a `ServerKey`. I will correct the message.